### PR TITLE
pump: check txn TTL before check IsCommitted

### LINKED ISSUE #1024: pump: lost transaction when ttl exceeded
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

This bug occurs when commit duration is more than 10min.

Use these patches for easy reproducing.

tidb(upstream): https://github.com/you06/tidb/tree/reproduce-slow-commit
binlog(pump): https://github.com/you06/tidb-binlog/tree/reproduce-slow-commit

From upstream, run the following statement

```sql
create table t(id int primary key, val int);
begin;
insert into t values(1, 1);
commit; -- this will block 20 seconds
```

Check downstream.

2. What did you expect to see?

```
MySQL [test]> select * from t;
+----+------+
| id | val  |
+----+------+
|  1 |    1 |
+----+------+
1 row in set (0.044 sec)
```

3. What did you see instead?

```
MySQL [test]> select * from t;
Empty set (0.045 sec)
```

pump's log

```log
[2020/12/08 07:34:04.263 +00:00] [INFO] [storage.go:541] ["known txn is committed from tikv"] ["start ts"=421373629621600257] ["commit ts"=0] [isDDL=false]
```

4. Please provide the relate downstream type and version of drainer.
(run `drainer -V` in terminal to get drainer's version)

```
Release Version: v4.0.8-3-g83984f4
Git Commit Hash: 83984f4a49898f6120157e240ca6a666ddeeafde
Build TS: 2020-11-27 04:55:57
Go Version: go1.15.2
Go OS/Arch: linux/amd64
```

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

fix #1024

### What is changed and how it works?

- check `status.TTL() > 0` before check `status.IsCommitted()`. if `status.TTL() > 0`, it means the txn's lock is still valid (neither commit nor rollback)

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
   - add `sleep` before `commit` in TiDB as the patch in #1024 (but do not modify `MaxTxnTTL`)
   - check no data lost in the downstream

Code changes

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### Release note
- Fix the problem that the data of long-time transactions may be lost

### CODE REVIEW

**Comment by csuzhangxc** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Comment by csuzhangxc** [comment]
@you06 @cfzjywxk @july2993 PTAL

File: pump/storage/storage.go [hunk 1]
```diff
@@ -480,6 +480,7 @@ func (a *Append) resolve(startTS int64) bool {
 	maxSecond := oracle.ExtractPhysical(uint64(latestTS)) / int64(time.Second/time.Millisecond)
 	elapseSecond := maxSecond - startSecond
 
+	// `GetTxnStatus` will not cleanup any locks now, but we still keep this logic and only `GetTxnStatus` after `maxTxnTimeoutSecond`.
```
Comment (cfzjywxk) [inline pump/storage/storage.go:483]: `GetTxnStatus` may cleanup expired locks whose `lock.startTS` + `lock.ttl` < `currentTS`. The returned `TxnStatus` is more complex than before we may refer the TiDB usage like https://github.com/pingcap/tidb/blob/master/store/tikv/lock_resolver.go#L361, or review the code in tikv https://github.com/tikv/tikv/blob/release-4.0/src/storage/mvcc/txn.rs#L905.

File: pump/storage/storage.go [hunk 2]
```diff
@@ -489,13 +490,24 @@ func (a *Append) resolve(startTS int64) bool {
...
+
+	log.Info("got commit status for unknown txn",
+		zap.Int64("start ts", startTS),
+		zap.Uint64("commit ts", status.CommitTS()),
+		zap.Uint64("ttl", status.TTL()),
+		zap.Reflect("action", status.Action()),
+		zap.Bool("isDDL", pbinlog.GetDdlJobId() > 0))
+
+	// check TTL (whether the lock is valid)
+	if status.TTL() > 0 {
```
Comment (cfzjywxk) [inline pump/storage/storage.go:509]: We could use the logic like this so we won't miss extra path.
```
if if status.TTL() > 0 {
    // lock is still valid.
} else {
    // The transaction status is certain now.
    if if status.IsCommitted() {
    } else{
    }
}
```

Comment (csuzhangxc) [reply pump/storage/storage.go:509]: 🤔 I don't know what's your `miss extra path` means?

File: pump/storage/storage.go (hunk 1 above)
Comment (csuzhangxc) [reply pump/storage/storage.go:483]: I want to express "do not abort txn now", 😹 .

"cleanup expired locks" will not take any side effects, is it right?

Comment (cfzjywxk) [reply pump/storage/storage.go:483]: "cleanup expired locks" is the way to abort the transaction. If the lock is rolled back, the transaction coult not commit any more.

File: pump/storage/storage.go (hunk 2 above)
Comment (cfzjywxk) [reply pump/storage/storage.go:509]: Just some usage format, it does not matter much.

**Comment by csuzhangxc** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Comment by cfzjywxk** [comment]
@you06 @MyonKeminta PTAL

**Comment by you06** [comment]
May update TiDB to this PR pingcap/tidb#21689

**Comment by csuzhangxc** [comment]
> May update TiDB to this PR [pingcap/tidb#21689](https://github.com/pingcap/tidb/pull/21689)

done in https://github.com/pingcap/tidb-binlog/pull/1025/commits/80c7c39639c38135326f33ff2d39c5b35d04a3d5.

**Comment by csuzhangxc** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0

**Review by you06 — COMMENTED** [review]
LGTM

**Comment by overvenus** [comment]
LGTM

**Review by overvenus — APPROVED** [review]

**Review by july2993 — APPROVED** [review]

**Pull request merged**